### PR TITLE
Fix confidence threshold cache invalidation and filtering logic

### PR DESCRIPTION
**Pull request by eugene123tw**
### Summary

Fixing this issue: https://github.com/open-edge-platform/geti/issues/940

🐛 Problem
`best_confidence_threshold` property in detection and instance segmentation models had two major issues:
* Cache invalidation bug: The property cached `_best_confidence_threshold` on first access, but when `hparams["best_confidence_threshold"]` was updated during validation, the cached value (`_best_confidence_threshold`) was never invalidated, causing stale thresholds to be used.

* Incorrect filtering location: Confidence threshold filtering was applied in validation_step() and test_step(), which meant:
  * Validation metrics were computed on pre-filtered data 
  * FMeasure couldn't properly compute optimal thresholds from the full prediction distribution 
  * This created a chicken-and-egg problem where the metric needed unfiltered data to find the best threshold

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/open-edge-platform/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/open-edge-platform/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/open-edge-platform/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/backend/native/models/instance_segmentation/base.py [hunk 1]
```diff
@@ -301,9 +377,37 @@ def _log_metrics(self, meter: Metric, key: Literal["val", "test"], **compute_kwa
                 {"best_confidence_threshold": best_confidence_threshold} if best_confidence_threshold else {}
             )
 
-            return super()._log_metrics(meter, key, **compute_kwargs)
-
-        raise ValueError(key)
+            super()._log_metrics(meter, key, **compute_kwargs)
+
+    def _filter_outputs_by_threshold(self, outputs: OTXPredBatch) -> OTXPredBatch:
```
Comment (kprokofi) [inline src/otx/backend/native/models/instance_segmentation/base.py:373]: Didn't we filter outputs by threshold before?

File: src/otx/backend/native/models/instance_segmentation/base.py [hunk 2]
```diff
@@ -264,35 +271,104 @@ def _export_parameters(self) -> TaskLevelExportParameters:
...
+            elif isinstance(meter, FMeasure):
+                fmeasure = meter
+
+            if fmeasure is not None and hasattr(fmeasure, "best_confidence_threshold"):
+                # Instead of using only the last best_confidence_threshold,
+                # we now average the last N values to get a more stable estimate,
+                # especially helpful for small datasets where it can vary greatly across epochs.
+                if "best_confidence_threshold_list" not in self.hparams:
+                    self.hparams["best_confidence_threshold_list"] = []
+                self.hparams["best_confidence_threshold_list"].append(fmeasure.best_confidence_threshold)
```
Comment (kprokofi) [inline src/otx/backend/native/models/instance_segmentation/base.py]: Could you explain why do we need `best_confidence_threshold_list`? Do you have some experiments to prove necessity of this average threshold? Also, could we avoid storing this list in hparams and save there only final one?

Comment (eugene123tw) [reply src/otx/backend/native/models/instance_segmentation/base.py]: This is explained in the comments.

Instead of relying on just the last best_confidence_threshold, we now average the last N values to obtain a more stable estimate. This is particularly useful for small datasets, where the threshold can fluctuate significantly between epochs.

The issue originates from: `https://github.com/open-edge-platform/geti/issues/940`.

File: src/otx/backend/native/models/instance_segmentation/base.py (hunk 1 above)
Comment (eugene123tw) [reply src/otx/backend/native/models/instance_segmentation/base.py:373]: we do, but the ordering of metric.compute and filtering is wrong.

File: src/otx/backend/native/models/base.py [hunk 3]
```diff
@@ -409,6 +409,9 @@ def on_save_checkpoint(self, checkpoint: dict[str, Any]) -> None:
             compiled_state_dict = checkpoint["state_dict"]
             checkpoint["state_dict"] = remove_state_dict_prefix(compiled_state_dict, "_orig_mod.")
         super().on_save_checkpoint(checkpoint)
+
+        if "best_confidence_threshold_list" in checkpoint["hyper_parameters"]:
+            checkpoint["hyper_parameters"].pop("best_confidence_threshold_list")
```
Comment (kprokofi) [inline src/otx/backend/native/models/base.py]: checkpoint["hyper_parameters"].pop("best_confidence_threshold_list", None) -> no need to check key

**Review by kprokofi — COMMENTED** [review]
Could you please update the Changelog? Seems like a new feature - heuristic. Worth to mention

File: src/otx/backend/native/models/instance_segmentation/base.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/backend/native/models/instance_segmentation/base.py]: I see, how much does it increase the f1-score estimation?

**Comment by eugene123tw** [comment]
@kprokofi 
I’ve decided to keep this PR simple and not implement averaging over `best_confidence_threshold_list` for now.

### Reasoning:
I experimented with the mean-based approach, but test scores still frequently drop to 0%. The core issue is high variance in confidence thresholds across validation epochs.

Here’s an example from 10 evaluation cycles: `[0.175, 0.1, 0.075, 0.075, 0.275, 0.2, 0.175, 0.675, 0.125, 0.45]`
* Mean: 0.234
* Std: 0.193
* CV (Coefficient of Variation): 82% → Extremely high variance

This level of variance makes the mean unreliable, and outliers like 0.675 significantly skew the result. As a result, the averaged threshold often doesn’t correspond to a good-performing value.

### Proposal:
Let’s address robust averaging separately in a future PR.
Proper handling would require:
* Coefficient of Variation analysis
* Median-based thresholdding for moderate variance
* Fallback/default thresholds for high variance
* Minimum evaluation size checks

However, these additions introduce a lot of complexity, which is out of scope for a bug-fix PR.

### So for this PR I will include:
* Fix the caching bug in best_confidence_threshold
* Ensure threshold updates only on best F1 scores
*  Skip threshold list averaging logic

### Next Steps:
Let’s revisit the robust thresholdding logic in a dedicated feature PR, where we can design, test, and validate the statistical approach more thoroughly.

**Review by kprokofi — APPROVED** [review]
Make sense, thank you for investigation!
